### ISSUE TITLE
Pyright discards type information with typeguard use.

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#1995]
**Describe the bug**
TypeGuards discard known type information and do not use it at all in the negative case.

**To Reproduce**

Have two mutually exclusive types in a Union and attempt to narrow them using a simple check.

**Expected behavior**

I expected that the typeguard use would retain existing known information and further narrow the type.

**Screenshots or Code**

```py
import asyncio
import inspect
from typing import Callable, Coroutine, Union, Any, TYPE_CHECKING
from typing_extensions import TypeGuard


if not TYPE_CHECKING:
    def reveal_type(_):
        ...

def my_typeguard(obj: Callable) -> TypeGuard[Callable[..., Coroutine]]:
    """ This was discovered prior to a PR that has not happened yet tp the typeshed as a result of it not fixing the issue """
    return inspect.iscoroutinefunction(obj)


def handle_cb(number: int, cb: Callable[[int], Union[Coroutine[None, None, int], int]]):
    if my_typeguard(cb):
        reveal_type(cb)
        asyncio.create_task(cb(number))
    else:
        reveal_type(cb)
        cb(number)
```

Code provided along side screenshot with diagnostic info inline

![image](https://user-images.githubusercontent.com/22035876/122308814-1e831980-cedb-11eb-855f-1fdafe531661.png)

The expectations for the first one is that it be narrowed to `Callable[[int], Coroutine[None, None, int]]`
The expectation for the second one is that since it *can't* be the first one, it is narrowed to `Callable[[int], int]`

Instead in the first case it throws away existing type information, and in the second it doesn't augment by ruling out anything.

**VS Code extension or command-line**

I can track down the specific version or verify it happens independently of what's bundled with the language server for VSCode as well if needed, but it's being provided via whichever pyright version was iincluded with `ms-python.python` v2021.6.944021595

**Additional context**

This is a slightly simplified version. Admittedly, in this specific simplified version, there is enough information.

This may also be a case for paramspec use in typeguards, but it was unexpected that this threw away existing known information as typeguard use would not be needed if that was my goal (typing.cast already exists)

**Comment by erictraut** [comment]
This is by design. Refer to [PEP 647](https://www.python.org/dev/peps/pep-0647/). It specifically says:

> The type is not narrowed in the negative case.

This was discussed at length during the development of TypeGuard. There are advantages and disadvantages of supporting type narrowing in the negative case. In the end, we decided that the disadvantages outweighed the advantages and decided against supporting type narrowing in this case.

**Comment by erictraut** [comment]
Interestingly, the design for the TypeGuard feature was inspired in part by TypeScript. Its implementation does support type narrowing in the negative case. This "feature" has bitten me many times. After discussing it with the TypeScript development team, they admitted that it was probably a mistake to support narrowing in the negative case. That, in part, informed our decision in PEP 647.

**Comment by Zomatree** [comment]
> There are advantages and disadvantages of supporting type narrowing in the negative case. In the end, we decided that the disadvantages outweighed the advantages and decided against supporting type narrowing in this case.

what disadvantages are there to supporting this? having this feature makes typing certain code a lot easier and i dont see any reason to not narrow down the type.

Also slightly related but `isinstance` does do type narrowing in the negative case, if instance does it i see dont see why typeguard cant.
![image](https://user-images.githubusercontent.com/39768508/122315325-ba2f7c80-cf11-11eb-8edd-2c6f18818a57.png)

**Comment by mikeshardmind** [comment]
> The type is not narrowed in the negative case.

It's not narrowing in the positive case either here though.

**Comment by erictraut** [comment]
Yes, the semantics of `isinstance` are well understood and documented, so type narrowing is safe to use in that case. 

For user-defined type guards, the semantics are not known because they're based on custom logic provided by the code.

> It's not narrowing in the positive case either here though.

Yes, it's narrowing in the positive case. The original type of `cb` is declared as `Callable[[int], Union[Coroutine[None, None, int], int]]`. In the positive branch of the test (the `if` clause), its type is narrowed to `Callable[..., Coroutine]` as dictated by the type guard. In the negative branch (the `else` clause) the type is left in its original unnarrowed type.

**Comment by mikeshardmind** [comment]
I'd disagree with calling that narrowing as it's discarding info already known, but I'll see if this can be rectified by updating types in the typeshed with paramspec use.

**Comment by erictraut** [comment]
Yes, we stretched the definition of "narrowing" in the context of PEP 647. Normally narrowing would require that the result is a subtype of the unnarrowed type, but we decided that would be too constraining for Python.

The TypeGuard mechanism works with generic functions (including the use of TypeVar and ParamSpec), so it gives you the flexibility to retain as much or as little of the original type as you want when defining a user-defined type guard function.

**Comment by mikeshardmind** [comment]
I wish I had had the time to follow the ParamSpect/TypeGuard discussions more closely as they were happening. This was a gigantic surprise to me because I saw the use of narrowing, and it just frankly doesn't make sense to how logical branching happens in real code.

If I already know something is either complex or a float, I don't suddenly forget that by checking if it had a non-real component. 
While all the tools are technically here, it seems to be harder to use than necessary for no tangible reason.

**Comment by erictraut** [comment]
Type narrowing is typically the process of determining a more specific type than the "wider" type you start with.

I'm not sure what you mean by "forgetting". Can you provide a concrete example?

Are you expecting a type checker to somehow know how to "merge" type arguments from the input type into the return type of the type guard function? That's not really possible in the general case. Or are you expecting it to create a union of the original and the type guard return type? That would defeat the purpose of the type guard.

**Comment by mikeshardmind** [comment]
the concrete example above works, but I can provide a simpler one if needed.

In the example above, only one portion of the union was compatible with a form of type guard. It should have been able to determine that if we already have info that it's in that union and only one member of the union was compatible, it's that member.

It's not an obvious gotcha here because there's no obvious use-case to me to discard that info that isn't better served by the existing `typing.cast` I'm fine with handling it with a paramspec, just a bit annoyed that this is by design.

The union above was essentially "provide a function that takes an int and returns an int. It may optionally be a coroutine function if the information is not available locally to not block the event loop" (the actual case is more complex, but the signature is known, just allows for it to optionally be async compatible)

inspect.iscoroutinefunction is used to narrow only the question of it was a coroutine function, but this method can't actually be made generic with paramspec in the typeshed as it is possible to pass non-functions to this (and get back `False`, it's not a misuse\*)

So I'll need a wrapper on my end anyhow, which might as well just return back which concrete type it is rather than bothering with making it generic.

While I can separate the logic, the actual place this comes into play involves decorators and in the interest of clean code with an obvious purpose, it seemed better to handle both cases from the decorator's side once, when decorating.


\* Demoing why this can't easily be made generic in the typeshed with this being as designed.
```py
>>> inspect.iscoroutinefunction(object())
False
```

**Comment by erictraut** [comment]
Your example above doesn't involve a union of two types. It involves a single type: `Callable[[int], Union[Coroutine[None, None, int], int]]`. One of the type arguments (the one corresponding to the return type of the callable) happens to be a union, but that's very different than a union of two types.

A type analyzer cannot simply "remember" some type arguments from one type and selectively apply them to another type. Type variables exist in the type system so these replacements can be specified in logic.

In your example above, the type guard function indicates that the "narrowed" type should be `Callable[..., Coroutine]`. This is equivalent to `(*args: Any, **kwargs: Any) -> Coroutine[Any, Any, Any]`. The `...` type argument in a `Callable` type indicates input parameters `(*args: Any, **kwargs: Any)`. The `Any` type arguments to `Coroutine` were not provided explicitly, but they are implicit by the rules spelled out in PEP 484.

If I understand you correctly, you are expecting the type analyzer to recognize that the input argument's type happens to conform to the general "shape" of `(*args: Any, **kwargs: Any) -> Coroutine[Any, Any, Any]` and replace the `...` and `Any` selectively with type arguments from the input type???? How would this work? Under what circumstances would it apply this replacement? Would it do it only for `Any` types but leave any others untouched? What if the input type doesn't fit the "shape" of the type guard return type — e.g. the input type is `Dict[int, int]` and the output is `Iterable[Any]`? I just don't see how this would work.

In any case, I'm glad you found a solution to the problem you are trying to solve.

**Comment by mikeshardmind** [comment]
> Your example above doesn't involve a union of two types. It involves a single type: Callable[[int], Union[Coroutine[None, None, int], int]]. One of the type arguments (the one corresponding to the return type of the callable) happens to be a union, but that's very different than a union of two types.

Conceptually, I don't see a meaningful distinction between

```
Callable[[int], Union[Coroutine[None, None, int], int]]
```
And 
```
Union[ Callable[[int], Coroutine[None, None, int]], Callable[[int], int] ]
```

These both describe the exact same Union of types, **the placement of the union does not change what types are valid here.**

Even if as a matter of practicality they are treated differently, the latter form also does not narrow to the compatible type in `TypeGuard[Coroutine]` if used above.

> If I understand you correctly, you are expecting the type analyzer to recognize that the input argument's type happens to conform to the general "shape" of (*args: Any, **kwargs: Any) -> Coroutine[Any, Any, Any] and replace the ... and Any selectively with type arguments from the input type???? How would this work?

Not exactly, but I do expect to not have to fight with tooling over something that a reasonable person could easily infer for themselves. It would essentially be more useful if typeguards were treated as additional information and not necessarily the concrete type.

With `Any` being allowed to be any type, I do think it's relatively reasonable to want to merge known type information to narrow, there's no case where this should be harmful, as the defaults for generic classes *already* behave in a similar manner.


As to **How would this work?**

Perhaps at this point, it's too late and there needs to be another PEP for a completely different mechanism to strictly narrow rather than replace type information.

It's rather simple logic to follow

1. A is an unknown element from the set {X, Y}
2. X ∈ B
3. Y ∉ B

From this, 

If A ∈ B, then A is X otherwise A is Y
If A ∉ B then A is Y otherwise A is X

(In plain English, if A is either X or Y and there is a set B which X is a part of but Y is not, a membership test for B is enough to disambiguate A to either X or Y)

This can even be extended further, including with partially unknown type information from just the theory.

1. A is an unknown element from the set {X, Y, Z}
2. X ∈ B
3. Y ∉ B
4. It is unknown whether Z ∉ B

If A ∈ B, then A is X or Z otherwise A is Y or Z
If A ∉ B then A is Y or Z otherwise A is X or Z

(In plain English, if A is either X, Y or Z and there is a set B which X is a part of Y is not, and Z cannot be determined, then a membership test for B can reduce the possible types of A to either {X, Z} or {Y, Z})


## What next

I'll see about detailing my thoughts and extracting the core essence of why this is useful to real code with concrete examples at a later point in time. I won't promise that I'll put in the time on it as right now my time is rather stretched thin, and most of the python code I own is being made less dynamic as it's the easier option, even if there's no reason it should need to be. 

The typeshed can use the solution above with ParamSpec + TypeVars (with some modification) and at least this specific case is already coverable.

## Note on PEP 647

While it is **entirely my fault for misunderstanding** the intended use of TypeGuard by skimming it while it was still a draft and missing certain details, I'm disappointed to see examples in the motivation section that can not be replicated with the intended solution.


---

I hope it's clear that my frustration is not with you Eric, I appreciate the time you've put into this and the further discussion about this. Hopefully, I can find some time to better express why this is useful and possibly prototype an actual implementation. Even though this example is "trivial" from a small bit of logical reasoning about sets of types, that doesn't mean that any current type checkers are necessarily designed with this in mind and it's something that would need some justification to add that sort of logic.
